tests/test_formatter.py

```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
     actual_problem_str = problem_to_string(expected_problem_obj)
     actual_problem_obj = problem_parser(actual_problem_str)
     assert actual_problem_obj == expected_problem_obj
+
+
+def test_typed_constants_formatting_in_domain() -> None:
+    """Test that types and typed constants are formatted correctly."""
+    t1, t2, t3 = "type_1", "type_2", "type_3"
+
+    a1, b1, c1 = constants("a b c", type_=t1)
+    d2, e2, f2 = constants("d e f", type_=t2)
+    g3, h3, i3 = constants("g h i", type_=t3)
+    j, k, lc = constants("j k l", type_=None)
+
+    # define the domain object.
+    domain = Domain(
+        "my_domain",
+        requirements=[Requirements.TYPING],
+        types={t1: None, t2: t1, t3: t1},
+        constants=[a1, b1, c1, d2, e2, f2, g3, h3, i3, j, k, lc],
+    )
+
+    domain_str = domain_to_string(domain)
+
+    assert domain_str == dedent(
+        """\
+    (define (domain my_domain)
+        (:requirements :typing)
+        (:types type_2 type_3 - type_1 type_1)
+        (:constants a b c - type_1 d e f - type_2 g h i - type_3 j k l)
+        (:predicates )
+    )"""
+    )
+
+
+def test_typed_objects_formatting_in_problem() -> None:
+    """Test that typed objects are formatted correctly."""
+    t1, t2, t3 = "type_1", "type_2", "type_3"
+
+    a1, b1, c1 = constants("a b c", type_=t1)
+    d2, e2, f2 = constants("d e f", type_=t2)
+    g3, h3, i3 = constants("g h i", type_=t3)
+    j, k, lc = constants("j k l", type_=None)
+
+    problem = Problem(
+        "problem-1",
+        domain_name="my_domain",
+        requirements=[Requirements.TYPING],
+        objects=[a1, b1, c1, d2, e2, f2, g3, h3, i3, j, k, lc],
+    )
+    problem_str = problem_to_string(problem)
+
+    assert problem_str == dedent(
+        """\
+    (define (problem problem-1)
+        (:domain my_domain)
+        (:requirements :typing)
+        (:objects a b c - type_1 d e f - type_2 g h i - type_3 j k l)
+        (:init )
+        (:goal (and))
```

Comment:
we should decide what to do with empty init and goal sections. The PDDL grammar specifies these sections are mandatory. We will tackle this in a next PR.